src/rest_vol.c

```diff
@@ -1257,10 +1256,22 @@ H5_rest_curl_read_data_callback(char *buffer, size_t size, size_t nmemb, void *i
     size_t       max_buf_size = size * nmemb;
     size_t       data_size = 0;
 
+    /* If all bytes sent, indicate transfer is finished */
+    if (uinfo->bytes_sent >= uinfo->buffer_size) {
+        return 0;
+    }
+
     if (inptr) {
```

Comment:
Instead of checking for buffer overrun each time, I feel like it'd be cleaner and more readable to update this code like:

    if (inptr) {
        size_t bytes_left;

        /* If all bytes sent, indicate transfer is finished */
        if (uinfo->bytes_sent >= uinfo->buffer_size) {
            return 0;
        }

        bytes_left = uinfo->buffer_size - uinfo->bytes_sent;
        data_size = (bytes_left > max_buf_size) ? max_buf_size : bytes_left;

        memcpy(buffer, uinfo->buffer + uinfo->bytes_sent, data_size);

        uinfo->bytes_sent += data_size;
    }

    return data_size;

We could add an assert in the code to catch buffer overruns in debug builds, but otherwise we shouldn't need to check for that case as long as the math adds up.